made solveset to work with eq having Max(),Min()

ishanaj opened the pull request:
Partial fix to issue #10158.
As mentioned in the issue ,equations having Max() and Min() will now be converted to Piecewise before
entering _solveset() via rewrite()-->(as implemented in PR #13309)

ping @smichr @asmeurer

ishanaj committed:
made solveset to work with eq having Max(),Min()

Equations having Max() and Min will now be converted to Piecewise before
entering _solveset() via rewrite()

ishanaj committed:
corrected failing tests

ylemkimon commented on sympy/solvers/solveset.py:
```diff
@@ -931,6 +932,9 @@ def solveset(f, symbol=None, domain=S.Complexes):
                 relationship between value and 0 is unknown: %s''' % b))
 
+    if f.has(Max) or f.has(Min):
+        f = f.rewrite(Piecewise())
```
You can just do `f = f.rewrite(Piecewise)`

ishanaj replied:
I tried the above change, but some of the tests in `sympy\solvers\tests\test_solveset.py` failed.

Failing tests:
`test_issue_9565`, `test_issue_10069`, `test_issue_9849`, `test_issue_9849 `, `test_invert_real`,`test_solve_abs`, `test_conditionset`, `test_issue_5132_2`

Looks like expressions with `abs()` donot work with this change.

ishanaj replied:
On running the test case individually, correct result is obtained. But, still the tests fail for this change. Can't get the reason why?
For [test_issue_9565()](https://github.com/sympy/sympy/blob/master/sympy/solvers/tests/test_solveset.py#L600)
```
>>>print(solveset(Abs((x - 1)/(x - 5)) <=  S(1)/3, x,S.Reals))         # test_issue_9565()
Interval(-1, 2) 
```
But while running tests `Type error` is raised: `TypeError: expecting Expr but got: ExprCondPair`
```
________________________________________________________________________________
_____________ sympy\solvers\tests\test_solveset.py:test_issue_9565 _____________
  File "d:\new\sympy\sympy\solvers\tests\test_solveset.py", line 601, in test_is                                                                                                                sue_9565
    assert print(solveset_real(Abs((x - 1)/(x - 5)) <= S(1)/3, x)) == Interval(-                                                                                                                1, 2)
  File "D:\new\sympy\sympy\solvers\solveset.py", line 957, in solveset_real
    return solveset(f, symbol, S.Reals)
  File "D:\new\sympy\sympy\solvers\solveset.py", line 948, in solveset
    f, symbol, domain=domain, relational=False)
  File "D:\new\sympy\sympy\solvers\inequalities.py", line 483, in solve_univaria                                                                                                                te_inequality
    period = periodicity(e, gen)
  File "D:\new\sympy\sympy\calculus\util.py", line 473, in periodicity
    g_s = decompogen(f, symbol)
  File "D:\new\sympy\sympy\solvers\decompogen.py", line 47, in decompogen
    result += [f.subs(f.args[0], symbol)] + decompogen(f.args[0], symbol)
  File "D:\new\sympy\sympy\solvers\decompogen.py", line 37, in decompogen
    raise TypeError('expecting Expr but got: `%s`' % func_name(f))
TypeError: expecting Expr but got: `ExprCondPair`
 
```

ylemkimon replied:
Do you mean changing `Piecewise` to `Piecewise()` or including other functions like `Abs`?

ishanaj replied:
@ylemkimon I meant including other functions(like `abs()`).
I had removed the `if ` condition. Then the tests failed

Previous code:
```
    else:
            raise NotImplementedError(filldedent('''
                relationship between value and 0 is unknown: %s''' % b))

    if f.has(Max) or f.has(Min):
        f = f.rewrite(Piecewise())

    if isinstance(f, Eq):
```
I tried by removing the `if` condition.
```
    else:
            raise NotImplementedError(filldedent('''
                relationship between value and 0 is unknown: %s''' % b))
 
    f = f.rewrite(Piecewise())

    if isinstance(f, Eq):
```
Is there any difference between `Piecewise` and `Piecewise()`? Although it doesn't seem to be while running this code. I am getting the same results for both.

ylemkimon replied:
`Abs` has its own solve method `_solve_abs`. So you can leave out `Abs`.

ylemkimon replied:
`Piecewise()` is an instance of `Piecewise` class, but `rewrites` automatically uses its class. So there is no difference in the result and an instantiation is unnecessary.

ishanaj committed:
changed 'Piecewise()' to 'Piecewise'

ishanaj commented:
@Yathartha22 Would it be better if we shift the line:`equation = equation.rewrite(Piecewise)`  inside the function `_solve_min_max()`?
Otherwise it looks fine to me.

Yathartha22 commented:
Sure, you can.

ishanaj committed:
added a new function _solve_min_max()

The rewrite statement is written inside the function _solve_min_max(),
which rewrites the eq(having Min or Max) into piecewise and further calls _solveset() and returns the result.
This function is called inside _solveset.

smichr commented:
Is this a good idea to create little helpers for things that will only rewrite themselves? Is it better to just attempt a `rewrite(Piecewise)` and `piecewise_fold` at the outset of solveset and then handle the Piecewise as its own entity?

Yathartha22 commented:
I guess you are right @smichr creating such small helper functions where only rewriting takes place, would just include unnecessary computation. I think simply rewriting eq to `Piecewise` at first will be better.

ishanaj committed:
removed function solve_min_max()

ishanaj commented:
@smichr Please review.

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -931,6 +932,9 @@ def solveset(f, symbol=None, domain=S.Complexes):
             raise NotImplementedError(filldedent('''
                 relationship between value and 0 is unknown: %s''' % b))
 
+    if f.has(Max) or f.has(Min):
```
Why not just let anything that can be rewritten as Piecewise benefit:

```diff
diff --git a/sympy/solvers/solveset.py b/sympy/solvers/solveset.py
index 1ce7e20..eaeff0c 100644
--- a/sympy/solvers/solveset.py
+++ b/sympy/solvers/solveset.py
@@ -19,6 +19,7 @@
 from sympy.functions import (log, Abs, tan, cot, sin, cos, sec, csc, exp,
                              acos, asin, acsc, asec, arg,
                              piecewise_fold)
+from sympy.functions.elementary.piecewise import Piecewise
 from sympy.functions.elementary.trigonometric import (TrigonometricFunction,
                                                       HyperbolicFunction)
 from sympy.functions.elementary.miscellaneous import real_root
@@ -995,6 +996,16 @@ def solveset(f, symbol=None, domain=S.Complexes):
             raise NotImplementedError(filldedent('''
                 relationship between value and 0 is unknown: %s''' % b))

+    if domain.is_subset(S.Reals):
+        if symbol.is_real is False:
+            return S.EmptySet
+        if symbol.is_real is None:
+            r = Dummy('r', real=True)
+            f = f.xreplace({symbol: r})
+            return solveset(f, r, domain).xreplace({r: symbol})
+    f = f.rewrite(Piecewise)
+
     if isinstance(f, Eq):
         from sympy.core import Add
         f = Add(f.lhs, - f.rhs, evaluate=False)
```
This gives
```
>>> solveset(x*Max(x, 15) - 10, x)
{2/3}
>>> solveset(abs(x)-1,x,S.Reals)
{-1, 1}
>>> solveset(sign(x)-1,x,S.Reals)
Interval.open(0, oo)
```

I'm not sure if the rewrite should occur only if the domain is real, however. If so, it should be part of the `if`-block.

ishanaj replied:
@smichr I think the above `if` condition is a good thing to be done.
But `test_issue_9565` and other tests(as I have mentioned above) fails. This is because of rewriting even the `Abs` to `Piecewise`. I think since `Abs` has its own `_solve_abs` equations with `Abs` should not be included in the rewriting part.

smichr commented on sympy/solvers/tests/test_solveset.py:
```diff
@@ -931,6 +931,8 @@ def test_solveset():
                                                   S.Integers)
     # issue 13825
     assert solveset(x**2 + f(0) + 1, x) == {-sqrt(-f(0) - 1), sqrt(-f(0) - 1)}
+    assert solveset(x*Max(x, 15) - 10, x) == FiniteSet(S("2/3"))                                #issue 10158
```
as in the case immediately above, just put the reference number on its own line

ishanaj replied:
Sure!

ishanaj committed:
Only eq with Abs are excluded from being rewritten into Piecewise

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -932,7 +932,7 @@ def solveset(f, symbol=None, domain=S.Complexes):
                 relationship between value and 0 is unknown: %s''' % b))
 
-    if f.has(Max) or f.has(Min):
+    if not f.has(Abs):
```
I think there are (at least) two more robust ways to do this:

1) mask off the Abs with dummies, rewrite, then re-substitute in the Abs
2) extract and rewrite only functions that have x in their free symbols

What do your current mods give for `Max(abs(x-3)-1,x+2)-3)`? The answer should be `[-1, 1]`.

smichr replied:
> give for 
I mean `solveset(..., x, S.Reals)`

ishanaj replied:
>What do your current mods give for Max(abs(x-3)-1,x+2)-3)? The answer should be [-1, 1].

It gives a `ConditionSet`.(same as the current master)
..And I am working on the two ways.

smichr replied:
I think the most robust way to go about it is to get the `[a for a in expr.atoms(Function) if x in a.free_symbols]` and rewrite those as Piecewise and those that changed get put into a replacement dictionary that is used to replace their occurrence in the original expression and then fold that expression and solve. Doing this will keep non-x dependent functions from being rewritten. e.g.

```
>>> solve(piecewise_fold(Abs(x-1).rewrite(Piecewise)-abs(y)),x)
[Abs(y) + 1, Piecewise((-Abs(y) + 1, -Abs(y) < 0), (nan, True))]
>>> solve(piecewise_fold((Abs(x-1)-abs(y)).rewrite(Piecewise)),x)
[Piecewise((-y + 1, y > 0), (nan, True)), Piecewise((-y + 1, y < 0), (nan, True)), 
Piecewise((y + 1, y >= 0), (nan, True)), Piecewise((y + 1, y < 0), (nan, True))]
```

ishanaj replied:
@smichr The above modification is just resolves problem with  the multivariate equations. I think the idea of masking off(as mentioned above) should also be implemented in addition to it. Else the `test_issue_9565` would fail as it fails when the `Abs` is converted to `Piecewise`.

ishanaj committed:
masked off Abs() before rewriting into Piecewise

ishanaj commented:
@smichr Please have a look.
Eq having `Abs()` are first masked off(i.e. `Abs()` part is replaced with a Dummy variable) and the changes are stored in `rep_dict` which is later used to replace back the masked off part. This is being done so as to prevent `Abs` from being rewritten into `Piecewise` as it might heckle the workflow of `_solve_abs()` (as in the case of `test_issue_9565`)

Although `test_issue_5132_2` fails, but I am getting the correct result while running it locally. Not able to get the reason.

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -931,6 +932,26 @@ def solveset(f, symbol=None, domain=S.Complexes):
+        e = [a for a in f.atoms(Abs) if symbol in a.free_symbols]
+        rep = numbered_symbols("t", cls=Dummy)
+        rep = [next(rep) for _ in range(len(e))]
+        newf = f
```
You can define a `reps` and `inv_reps` as:

```
reps = dict(zip(e, numered_symbols("t", cls=Dummy)))
inv_reps = [(v, k) for k, v in reps.items()]
```

Restoration should be done with `xreplace` since the keys will be symbols: `f = f.rewrite(Piecewise).xreplace(inv_reps)`

smichr replied:
for example...

```diff
diff --git a/sympy/solvers/solveset.py b/sympy/solvers/solveset.py
index 85b4d7d..9e22c9b 100644
--- a/sympy/solvers/solveset.py
+++ b/sympy/solvers/solveset.py
@@ -32,6 +32,7 @@
 from sympy.solvers.polysys import solve_poly_system
 from sympy.solvers.inequalities import solve_univariate_inequality
 from sympy.utilities import filldedent
+from sympy.utilities.iterables import numbered_symbols
 from sympy.calculus.util import periodicity, continuous_domain
 from sympy.core.compatibility import ordered, default_sort_key, is_sequence
 from sympy.functions.elementary.piecewise import Piecewise
@@ -932,25 +933,18 @@ def solveset(f, symbol=None, domain=S.Complexes):
             raise NotImplementedError(filldedent('''
                 relationship between value and 0 is unknown: %s''' % b))

-    rep = []
-    if f.has(Abs):
-        # masking off the Abs() part from being rewritten into Piecewise
-        from sympy import numbered_symbols
-        e = [a for a in f.atoms(Abs) if symbol in a.free_symbols]
-        rep = numbered_symbols("t", cls=Dummy)
-        rep = [next(rep) for _ in range(len(e))]
-        newf = f
-        rep_dict = {}
-        for r in e:
-            newf =  newf.replace(r, rep[e.index(r)])
-            rep_dict[rep[e.index(r)]] = r
-        f = newf
-
+    reps = dict(zip(
+        # do larger expressions first in case they contain
+        # other Abs
+        reversed(list(ordered(
+            [i for i in f.atoms(Abs) if symbol in i.free_symbols]))),
+        numbered_symbols("t", cls=Dummy)))
+    for k in reps:
+        f = f.replace(k, reps[k])
     f = f.rewrite(Piecewise)
-
-    for i in rep:
-        # replacing back the masked off part
-        f = f.replace(i, rep_dict[i])
+    if reps:
+        ireps = dict([(v, k) for k, v in reps.items()])
+        f = f.xreplace(ireps)

     if isinstance(f, Eq):
         from sympy.core import Add
```

ishanaj committed:
improved logic

ishanaj commented:
@smichr I have used `OrderedDict` for the  case of nested `Abs`.
Please have a look .
Also have added a test case.

smichr committed:
use list instead of OrderedDict

smichr commented:
Nice catch on the ordering...I further modified this to simply get rid of the dict and to use a list instead.

I also had the issue with the failure in tests that didn't fail locally. Did that get resolved by this change?

ishanaj committed:
corrected the way of replacement in accordance with the list

ishanaj committed:
Merge branch 'min_max' of https://github.com/ishanaj/sympy into min_max

ishanaj commented:
@smichr I have made the initial `replace` also to work in accordance with the list.
>Did that get resolved by this change?

No, the test still fails. I am working on it

smichr commented on sympy/solvers/solveset.py:
```diff
@@ -934,20 +933,19 @@ def solveset(f, symbol=None, domain=S.Complexes):
         numbered_symbols("t", cls=Dummy)))
     for k in reps:
-        f = f.replace(k, reps[k])
+        f = f.replace(k[0], k[1])
```
oops -- thanks.

I might also recommend that you either traverse the elements as:

```
for o, n in reps:
    f = f.replace(o, n)
```
or simply write `f = f.replace(*k)` which unpacks the tuple for you.

It also occurs to me that the arg of Abs may contain things that should be rewritten as Piecewise -- will that happen when the Abs is handled?

ishanaj replied:
@smichr For cases with Abs having args that should be rewritten into `Piecewise` like:
```
>>>solveset(Abs(Max(x, 2) +4*x)- 33, x, S.Reals)            # for this branch
Union(ConditionSet(x, Eq(4*x + Max(2, x) - 33, 0), S.Reals), \
ConditionSet(x, Eq(4*x + Max(2, x) + 33, 0), S.Reals))
# same as the current master
```
While the correct answer should be:
(which comes when it is rewitten into Piecewise)
```
>>>solveset(Abs(Max(x, 2) +4*x)- 33, x, S.Reals)   
{-35/4, 33/5}
```

smichr commented:
I recommend that you remove the `_solve_abs` routine and simply rewrite f unconditionally in terms of Piecewise, fold the result and let the solution process deal with that. If it can't solve the rewritten expression then that is a new issue for `solveset`.

smichr commented:
I believe the difference between the results when running the suite or running locally is whether the variables is declared as real or not. In the suite, for example, x is real. See if that works for you and helps you figure out the discrepancy.

smichr commented:
> remove the _solve_abs routine

I see why it is better to keep this.

ishanaj commented:
>I believe the difference between the results when running the suite ....

I tried declaring all the `symbols` with `real=True` but still the the same thing happens.

Nevertheless, I found the problem. In `test_issue_5132_2`, while solving the `nonlin` equation, solveset is called and inside solveset the unmasked `Abs` gets converted into `Piecewise`

f = `I*(2*_n*pi + arg(-z**2 + sin(y)))/2 - log(-z**2 + sin(y))/2 + log(Abs(z**2 - sin(y)))/2` 
gets converted to:
`I*(2*_n*pi + arg(-z**2 + sin(y)))/2 - log(-z**2 + sin(y))/2 + log(Piecewise((z**2 - sin(y), z**2 - sin(y) >= 0), (-z**2 + sin(y), True)))/2`
 where the `symbol` for the eq is dummy variable`_n`.

Because of `[i for i in f.atoms(Abs) if symbol in i.free_symbols]` in the commit only `Abs` with its args having `symbol` gets masked off, and the `Abs` in the eq gets converted into `Piecewise` which shows the error.

I think the condition should be reduced to:
```
reps = list(zip(
        # do larger expressions first in case they contain
        # other Abs
        reversed(list(ordered(
            [i for i in f.atoms(Abs)]))),
        numbered_symbols("t", cls=Dummy)))
```
so that every `Abs` gets masked off.
ping @smichr
